synapse/lib/stormtypes.py

```diff
@@ -1478,6 +1479,183 @@ async def _libIngest(self, name, data):
+        text = await tostr(filler)
+
+        if size < 1 or size > 10000:
+            mesg = '$lib.pipe.gen() size must be 1-10000'
+            raise s_exc.BadArg(mesg=mesg)
+
+        pipe = Pipe(self.runt, size)
+
+        varz = self.runt.vars.copy()
+        varz['pipe'] = pipe
+
+        opts = {'vars': varz}
+
+        query = await self.runt.getStormQuery(text)
+        runt = self.runt.getModRuntime(query, opts=opts)
+
+        async def coro():
+            try:
+                async for item in runt.execute():
+                    await asyncio.sleep(0)
+
+            except asyncio.CancelledError: # pragma: no cover
+                raise
+
+            except Exception as e:
+                await self.runt.warn(f'pipe filler error: {e}', log=False)
+
+            await pipe.close()
+
+        self.runt.snap.schedCoro(coro())
+
+        return pipe
+
+@registry.registerType
+class Pipe(StormType):
+    '''
+    A Storm Pipe provides fast ephemeral queues.
+    '''
+    def __init__(self, runt, size):
+        StormType.__init__(self)
+        self.runt = runt
+
+        self.locls.update(self.getObjLocals())
+        self.queue = s_queue.Queue(maxsize=size)
+
+    def getObjLocals(self):
+        return {
+            'put': self._methPipePut,
+            'puts': self._methPipePuts,
+            'slice': self._methPipeSlice,
+            'slices': self._methPipeSlices,
+            'size': self._methPipeSize,
+        }
+
+    async def _methPipePuts(self, items):
+        '''
+        Add a list of items to the Pipe.
+
+        Args:
+            items (list): A list of items to add.
+        '''
+        items = await toprim(items)
+        return await self.queue.puts(items)
+
+    async def _methPipePut(self, item):
+        '''
+        Add a single item to the Pipe.
+
+        Args:
+            item: An object to add to the Pipe.
+        '''
+        item = await toprim(item)
+        return await self.queue.put(item)
+
+    async def close(self):
+        '''
+        Close the pipe for writing.  This will cause
+        the slice()/slices() API to return once drained.
+        '''
+        await self.queue.close()
+
+    async def _methPipeSize(self):
+        '''
+        Retrieve the number of items in the Pipe.
+
+        Returns:
+            int: The number of items in the Pipe.
+        '''
+        return await self.queue.size()
+
+    async def _methPipeSlice(self, size=1000):
+        '''
+        Return a list of up to size items from the Pipe.
+
+        Args:
+            size: The max number of items to return (default 1000)
+
+        Returns:
+            list: A list of at least 1 item from the Pipe.
+        '''
+        size = await toint(size)
+        if size < 1 or size > 10000:
+            mesg = '$pipe.slice() size must be 1-10000'
+            raise s_exc.BadArg(mesg=mesg)
+
+        items = await self.queue.slice(size=size)
+        if items is None:
+            return None
```

Comment:
Would this be safer to return a List of zero objects? That way there are not type issues?

Reply:
None here signifies that the queue is closed and empty.  Most use cases will use `slices()` for minimal code